src/resources/tcp_esp32.cc

```diff
@@ -408,26 +384,44 @@ PRIMITIVE(read)  {
       return Smi::from(-1);
     }
 
-    ByteArray* array = allocate_read_buffer(process, p);
+    if (p->tot_len == 0) return Smi::from(-1);  // Nothing to read.
+
+    int offset = socket->read_offset();
```

Comment:
`socket->read_offset()` was just unused before? I might be inclined to couple the offset and buffer more.

```
int offset;
pbuf* p = socket->get_read_buffer(p, &offset);
```